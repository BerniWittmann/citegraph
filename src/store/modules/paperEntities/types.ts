import { PaperEntity } from '@/models/paperEntities'

export interface PaperEntitiesState {
  entityType: string | undefined,
  entities: Array<PaperEntity>,
  entityCount: number,
  activeEntity: PaperEntity | undefined
}
