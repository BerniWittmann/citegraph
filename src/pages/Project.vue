<template>
  <div class="project">
    <v-alert v-if="!hasProject" type="error" prominent>
      {{ $t('project.not_available') }}
    </v-alert>

    <div v-if="hasProject">
      <h1>{{ project.name }}</h1>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project from '@/models/project'

@Component
export default class ProjectPage extends Vue {
  get project (): Project | undefined {
    return this.$store.getters['projects/activeProject']
  }

  get hasProject (): boolean {
    return this.$store.getters['projects/hasActiveProject']
  }
}
</script>
