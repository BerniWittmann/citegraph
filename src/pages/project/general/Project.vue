import { ProjectStates } from '@/models/project'
<template>
  <div class="project">
    <div class="d-flex mr-5 mt-4">
      <h1>{{ project.name }}</h1>
      <v-spacer/>
      <v-btn :to="{ name: 'projects.single.edit', params: { projectId: project.id } }">
        <v-icon>mdi-pencil</v-icon>
        {{ $t('project.edit.button_text') }}
      </v-btn>
    </div>
    <v-container v-if="hasNoDataYet">
      <v-row justify="center">
        <v-col :lg="3" class="d-flex justify-center flex-column text-center">
          <empty-icon class="align-self-center mb-6"></empty-icon>
          <h3 class="headline">{{ $t('project.single.no_data.description') }}</h3>
          <p>{{ $t('project.single.no_data.import_data') }}</p>
          <v-btn color="primary" rounded large :to="{ name: 'projects.single.import', params: { projectId: project.id } }">
            {{ $t('project.single.no_data.button') }}
          </v-btn>
        </v-col>
      </v-row>
    </v-container>
    <v-container v-else>
      <v-row v-if="datasetInformation">
        <v-col>
          <v-card>
            <v-card-title>{{ $t('project.single.dataset.title') }}</v-card-title>
            <v-card-text>
              <v-container>
                <v-row>
                  <v-col v-for="(count, name) in datasetInformation" :key="name">
                    <v-card outlined class="d-flex flex-column justify-center text-center pa-5">
                      <p class="headline">{{count}}</p>
                      <p class="mb-0">{{ $t('project.single.dataset.' + name ) }}</p>
                    </v-card>
                  </v-col>
                </v-row>
              </v-container>
            </v-card-text>
          </v-card>
        </v-col>
      </v-row>
      <v-row>
        <v-col lg="3" md="6" cols="12" v-if="project.topAuthors">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_authors') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="author in project.topAuthors" :key="author.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'author',
                 entityId: author.id
                }}">
                  {{ author.firstName }} {{ author.lastName }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.topRecords">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.top_10_records') }}
            </v-card-title>
            <v-card-text>
              <v-list>
                <v-list-item v-for="record in project.topRecords" :key="record.id" :to="{ name: 'projects.single.explore.view', params: {
                 projectId: project.id,
                 queryByType: 'record',
                 entityId: record.id
                }}">
                  {{ record.title }}
                </v-list-item>
              </v-list>
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12" v-if="project.countryRelevanceData">
          <v-card>
            <v-card-title>
              {{ $t('project.single.overview.country_relevance') }}
            </v-card-title>
            <v-card-text>
              TODO Diagram here
            </v-card-text>
          </v-card>
        </v-col>
        <v-col lg="3" md="6" cols="12">
          <v-row no-gutters>
            <v-col v-if="project.topKeywords" class="mb-6">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.top_10_keywords') }}
                </v-card-title>
                <v-card-text>
                  <v-chip-group multiple column>
                    <v-chip v-for="keyword in project.topKeywords" :key="keyword">{{ keyword }}</v-chip>
                  </v-chip-group>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
          <v-row no-gutters>
            <v-col v-if="project.counts">
              <v-card>
                <v-card-title>
                  {{ $t('project.single.overview.visualizations') }}
                </v-card-title>
                <v-card-text>
                  <div class="d-flex flex-column justify-center text-center pa-5" v-if="hasNoVisualizationsYet">
                    <p class="mb-0">{{ $t('project.single.overview.no_visualizations_yet') }}</p>
                    <v-btn color="primary" rounded>{{ $t('project.single.overview.add_visualization') }}</v-btn>
                  </div>
                  <v-card outlined class="d-flex flex-column justify-center text-center pa-5" v-else>
                    <p class="headline">{{ project.counts.visualizations }}</p>
                    <p class="mb-0">{{ $t('project.single.overview.visualizations') }}</p>
                  </v-card>
                </v-card-text>
              </v-card>
            </v-col>
          </v-row>
        </v-col>
      </v-row>
    </v-container>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Project, { ProjectStates } from '@/models/project'
import EmptyIcon from '@/assets/svgs/Empty.vue'

@Component({
  components: {
    EmptyIcon
  }
})
export default class ProjectPage extends Vue {
  get project (): Project {
    return this.$store.getters['projects/activeProject']
  }

  get datasetInformation (): Object | undefined {
    if (!this.project.counts) return undefined
    const data = { ...this.project.counts }
    delete data['visualizations']
    return data
  }

  get hasNoDataYet (): boolean {
    return this.project.state === ProjectStates.IMPORT
  }

  get hasNoVisualizationsYet (): boolean {
    return this.project.state === ProjectStates.EXPLORE
  }
}
</script>
