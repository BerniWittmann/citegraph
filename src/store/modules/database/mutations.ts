import { DatabaseState } from './types'
import { MutationTree } from 'vuex'

export const mutations: MutationTree<DatabaseState> = {}
