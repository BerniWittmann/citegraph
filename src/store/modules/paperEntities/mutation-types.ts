export const SET_ENTITIES = 'SET_ENTITIES'
export const SET_ENTITY_TYPE = 'SET_ENTITY_TYPE'
export const SET_ENTITY_COUNT = 'SET_ENTITY_COUNT'
