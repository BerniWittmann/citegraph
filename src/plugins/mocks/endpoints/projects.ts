import { AxiosRequestConfig } from 'axios'
import MockAdapter from 'axios-mock-adapter'
import db from '../db'
import Project from '@/models/project'

const projects = db.getCollection('projects')

function getProjectId (config: AxiosRequestConfig): number {
  const urlParts: Array<string> = config.url!.split('/')
  return parseInt(urlParts[4])
}

export function setupMocks (mock: MockAdapter) {
  mock.onGet(/\/projects\/\d+/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    return [200, project]
  })

  mock.onGet('/projects').reply(() => {
    return [200, projects.find()]
  })

  mock.onPost('/projects').reply((config: AxiosRequestConfig) => {
    try {
      const projectName = JSON.parse(config.data).name.trim()
      const newProject = new Project({ id: projects.count() + 1, name: projectName })
      projects.insert(newProject)
      db.saveDatabase()
      return [201, newProject]
    } catch {
      return [400, 'Could not read body']
    }
  })

  mock.onDelete(/\/projects\/\d+/).reply((config: AxiosRequestConfig) => {
    const project: Resultset<Project> = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    projects.remove(project)
    db.saveDatabase()
    return [200, project]
  })

  mock.onPut(/\/projects\/\d+/).reply((config: AxiosRequestConfig) => {
    let project: Project = projects.findOne({ 'id': getProjectId(config) })
    if (!project) {
      return [404]
    }
    const projectName = JSON.parse(config.data).name.trim()
    project = projects.update({
      ...project,
      name: projectName
    })
    db.saveDatabase()
    return [200, project]
  })
}
