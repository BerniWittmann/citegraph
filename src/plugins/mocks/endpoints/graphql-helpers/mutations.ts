import { PaperEntityFields } from '@/models/paperEntities'
import { QueryConfiguration } from '@/plugins/mocks/endpoints/graphql-helpers/queries'
import db from '../../db'

type MutationArguments = {
  id?: string
}

function parseQuery (query: string): MutationArguments {
  const idString = query.match(/id: [a-zA-Z0-9]+/)
  const id = idString ? idString[0].toString().substring(4) : undefined
  return {
    id
  }
}

function parseData (query: string): Partial<PaperEntityFields> {
  // extract update data
  const result: Record<string, any> = {}
  const items = query.match(/\w+: ".*"/g) || []
  items.forEach((item: string) => {
    const key = item.split(': ')[0]
    const val = item.split(': ')[1]
    result[key] = val.substring(1, val.length - 1).replace(/\/"/g, '')
  })
  return result
}

function updateOne (queryParams: MutationArguments, data: Partial<PaperEntityFields>, configuration: QueryConfiguration) {
  let entity = configuration.collection.findOne({ id: queryParams.id })
  if (!entity) {
    throw new Error('Entity not found')
  }
  for (let key in data) {
    // @ts-ignore
    entity[key] = data[key]
  }
  return configuration.collection.update(entity)
}

export function createMutationResponse (query: string, configuration: QueryConfiguration) {
  const queryParams = parseQuery(query)
  if (!queryParams.id) {
    return [400, 'No id given']
  }
  try {
    const response = updateOne(queryParams, parseData(query), configuration)
    db.saveDatabase()
    return [200, {
      data: {
        [configuration.queryName]: {
          ...response
        }
      }
    }]
  } catch (err) {
    return [404, err]
  }
}
