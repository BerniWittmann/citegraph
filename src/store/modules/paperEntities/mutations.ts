import { PaperEntitiesState } from './types'
import { MutationTree } from 'vuex'
import {
  SET_ACTIVE_ENTITY,
  SET_ENTITIES,
  SET_ENTITY_COUNT,
  SET_ENTITY_TYPE
} from '@/store/modules/paperEntities/mutation-types'
import { PaperEntity } from '@/models/paperEntities'

export const mutations: MutationTree<PaperEntitiesState> = {
  [SET_ENTITIES] (state: PaperEntitiesState, payload: Array<PaperEntity>) {
    state.entities = payload
  },
  [SET_ENTITY_TYPE] (state: PaperEntitiesState, payload: string | undefined) {
    state.entityType = payload
  },
  [SET_ENTITY_COUNT] (state: PaperEntitiesState, payload: number) {
    state.entityCount = payload
  },
  [SET_ACTIVE_ENTITY] (state: PaperEntitiesState, payload: PaperEntity | undefined) {
    state.activeEntity = payload
  }
}
