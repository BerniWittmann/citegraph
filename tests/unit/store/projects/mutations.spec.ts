import { mutations } from '@/store/modules/projects/mutations'
import * as mutationTypes from '@/store/modules/projects/mutation-types'
import Project from '@/models/project'
import { ProjectsState } from '@/store/modules/projects/types'

describe('store/modules/projects/mutations', () => {
  it('provides the mutations', () => {
    expect(mutations).toMatchSnapshot()
  })

  describe('SET_ACTIVE_PROJECT', () => {
    it('sets the active Project', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: []
      }
      const payload: Project = new Project({ id: 1, name: 'My Project' })

      mutations[mutationTypes.SET_ACTIVE_PROJECT](state, payload)

      expect(state.activeProject).toEqual(payload)
    })

    it('overwrites an existing active project', () => {
      const state: ProjectsState = {
        activeProject: new Project({ id: 42, name: 'Other Project' }),
        projects: [],
        openProjects: []
      }
      const payload: Project = new Project({ id: 1, name: 'My Project' })

      mutations[mutationTypes.SET_ACTIVE_PROJECT](state, payload)

      expect(state.activeProject).toEqual(payload)
    })
  })

  describe('UNSET_ACTIVE_PROJECT', () => {
    it('unsets the active Project', () => {
      const state: ProjectsState = {
        activeProject: new Project({ id: 42, name: 'Other Project' }),
        projects: [],
        openProjects: []
      }

      mutations[mutationTypes.UNSET_ACTIVE_PROJECT](state)

      expect(state.activeProject).toBeUndefined()
    })

    it('also works when it is already unset', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: []
      }

      mutations[mutationTypes.UNSET_ACTIVE_PROJECT](state)

      expect(state.activeProject).toBeUndefined()
    })
  })

  describe('SET_PROJECTS', () => {
    it('saves the projects to the store', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: []
      }
      const projects: Array<Project> = [
        new Project({ id: 1, name: 'First Project' }),
        new Project({ id: 2, name: 'Second Project' })
      ]

      mutations[mutationTypes.SET_PROJECTS](state, projects)

      expect(state.projects).toEqual(projects)
    })

    it('overrides existing projects', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [
          new Project({ id: 99, name: 'Other Project' }),
          new Project({ id: 100, name: 'Another Project' })
        ],
        openProjects: []
      }
      const projects: Array<Project> = [
        new Project({ id: 1, name: 'First Project' }),
        new Project({ id: 2, name: 'Second Project' })
      ]

      mutations[mutationTypes.SET_PROJECTS](state, projects)

      expect(state.projects).toEqual(projects)
    })
  })

  describe('OPEN_PROJECT', () => {
    it('adds a new project to open projects', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: [
          new Project({ id: 1, name: 'First Project' }),
          new Project({ id: 2, name: 'Second Project' })
        ]
      }
      const project = new Project({ id: 42, name: 'My awesome Project' })

      mutations[mutationTypes.OPEN_PROJECT](state, project)

      expect(state.openProjects).toContainEqual(project)
      expect(state.openProjects.length).toEqual(3)
    })

    it('does not add a project twice', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: [
          new Project({ id: 1, name: 'First Project' }),
          new Project({ id: 2, name: 'Second Project' })
        ]
      }
      const project = new Project({ id: 2, name: 'Second Project' })

      mutations[mutationTypes.OPEN_PROJECT](state, project)

      expect(state.openProjects).toContainEqual(project)
      expect(state.openProjects.length).toEqual(2)
      expect(state.openProjects.map(project => project.id)).toEqual([1, 2])
    })

    it('can handle no open Projects before', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: []
      }
      const project = new Project({ id: 42, name: 'My awesome Project' })

      mutations[mutationTypes.OPEN_PROJECT](state, project)

      expect(state.openProjects).toContainEqual(project)
      expect(state.openProjects.length).toEqual(1)
    })
  })

  describe('CLOSE_PROJECT', () => {
    it('removes a project from open projects', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: [
          new Project({ id: 1, name: 'First Project' }),
          new Project({ id: 2, name: 'Second Project' })
        ]
      }
      const project = new Project({ id: 2, name: 'Second Project' })

      mutations[mutationTypes.CLOSE_PROJECT](state, project)

      expect(state.openProjects).not.toContainEqual(project)
      expect(state.openProjects.length).toEqual(1)
    })

    it('removes all instances of a project from open projects', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: [
          new Project({ id: 1, name: 'First Project' }),
          new Project({ id: 2, name: 'Second Project' }),
          new Project({ id: 2, name: 'Second Project' }),
          new Project({ id: 2, name: 'Second Project' })
        ]
      }
      const project = new Project({ id: 2, name: 'Second Project' })

      mutations[mutationTypes.CLOSE_PROJECT](state, project)

      expect(state.openProjects).not.toContainEqual(project)
      expect(state.openProjects.length).toEqual(1)
    })

    it('does not change open projects if project was not found', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: [
          new Project({ id: 1, name: 'First Project' }),
          new Project({ id: 2, name: 'Second Project' })
        ]
      }
      const project = new Project({ id: 42, name: 'Other Project' })

      mutations[mutationTypes.CLOSE_PROJECT](state, project)

      expect(state.openProjects).not.toContainEqual(project)
      expect(state.openProjects.length).toEqual(2)
      expect(state.openProjects.map(project => project.id)).toEqual([1, 2])
    })

    it('can handle an empty open projects array', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: []
      }
      const project = new Project({ id: 42, name: 'Other Project' })

      mutations[mutationTypes.CLOSE_PROJECT](state, project)

      expect(state.openProjects.length).toEqual(0)
    })
  })

  describe('ADD_PROJECT', () => {
    it('adds a project to the store', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [new Project({ id: 42, name: 'Other Project' })],
        openProjects: []
      }
      const project = new Project({ id: 1, name: 'My awesome Project' })

      mutations[mutationTypes.ADD_PROJECT](state, project)

      expect(state.projects.length).toEqual(2)
      expect(state.projects).toContain(project)
    })

    it('can add a project to an empty store', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: []
      }

      const project = new Project({ id: 1, name: 'My awesome Project' })

      mutations[mutationTypes.ADD_PROJECT](state, project)

      expect(state.projects.length).toEqual(1)
      expect(state.projects).toContain(project)
    })
  })

  describe('DELETE_PROJECT', () => {
    it('removes a project from the projects', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [
          new Project({ id: 1, name: 'First Project' }),
          new Project({ id: 2, name: 'Second Project' })
        ],
        openProjects: []
      }
      const project = new Project({ id: 2, name: 'Second Project' })

      mutations[mutationTypes.DELETE_PROJECT](state, project)

      expect(state.projects).not.toContainEqual(project)
      expect(state.projects.length).toEqual(1)
    })

    it('removes all instances of a project from the projects', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [
          new Project({ id: 1, name: 'First Project' }),
          new Project({ id: 2, name: 'Second Project' }),
          new Project({ id: 2, name: 'Second Project' }),
          new Project({ id: 2, name: 'Second Project' })
        ],
        openProjects: []
      }
      const project = new Project({ id: 2, name: 'Second Project' })

      mutations[mutationTypes.DELETE_PROJECT](state, project)

      expect(state.projects).not.toContainEqual(project)
      expect(state.projects.length).toEqual(1)
    })

    it('does not change projects if project was not found', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [
          new Project({ id: 1, name: 'First Project' }),
          new Project({ id: 2, name: 'Second Project' })
        ],
        openProjects: []
      }
      const project = new Project({ id: 42, name: 'Other Project' })

      mutations[mutationTypes.DELETE_PROJECT](state, project)

      expect(state.projects).not.toContainEqual(project)
      expect(state.projects.length).toEqual(2)
      expect(state.projects.map(project => project.id)).toEqual([1, 2])
    })

    it('can handle an empty open projects array', () => {
      const state: ProjectsState = {
        activeProject: undefined,
        projects: [],
        openProjects: []
      }
      const project = new Project({ id: 42, name: 'Other Project' })

      mutations[mutationTypes.DELETE_PROJECT](state, project)

      expect(state.projects.length).toEqual(0)
    })
  })
})
