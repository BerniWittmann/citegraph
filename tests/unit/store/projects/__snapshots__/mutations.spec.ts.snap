// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/modules/projects/mutations provides the mutations 1`] = `
Object {
  "CLOSE_PROJECT": [Function],
  "OPEN_PROJECT": [Function],
  "SET_ACTIVE_PROJECT": [Function],
  "SET_PROJECTS": [Function],
  "UNSET_ACTIVE_PROJECT": [Function],
}
`;
